### ISSUE TITLE
Can I add language support to harper at runtime?

### ISSUE THREAD

**Issue by andremarianiello** [Automattic/harper#331]
If I wanted to use harper on Haskell files, is there a way to integrate harper with the tree-sitter-haskell grammar at runtime? Perhaps I could point harper to my installation of the parser for dynamic loading? If not, what is the process of getting language support statically added to harper?

**Comment by grantlemons** [comment]
I don't think it's possible to integrate it at runtime, but adding support for parsing comments should be pretty trivial since it is just parsing comments.

**Comment by elijah-potter** [comment]
Hey @andremarianiello, we've added very rough initial support for Haskell. Give it whirl and let us know if there are any issues. 

@grantlemons, would you mind looking into whether structured comment support [as described here](https://writewithharper.com/docs/integrations/visual-studio-code) is applicable to Haskell?

**Comment by grantlemons** [comment]
> Hey [@andremarianiello](https://github.com/andremarianiello), we've added very rough initial support for Haskell. Give it whirl and let us know if there are any issues.
> 
> [@grantlemons](https://github.com/grantlemons), would you mind looking into whether structured comment support [as described here](https://writewithharper.com/docs/integrations/visual-studio-code) is applicable to Haskell?

Are you sure this is the right link? I don't see anything about structured comment support.

**Comment by andremarianiello** [comment]
My very basic test with Haskell worked nicely, and will definitely be useful going forward in my Haskell projects. Unfortunately I wasn't specific enough in my initial description, because I initially chose harper to help me grammar check blog posts that I write with _literate_ haskell. As expected, I get no hints from harper on literate haskell (.lhs) files. Would you accept a PR that adds literate haskell support? I saw it was removed in https://github.com/grantlemons/harper/commit/c100c6d9955c108526281069340a9579751481a8

**Comment by grantlemons** [comment]
> My very basic test with Haskell worked nicely, and will definitely be useful going forward in my Haskell projects. Unfortunately I wasn't specific enough in my initial description, because I initially chose harper to help me grammar check blog posts that I write with _literate_ haskell. As expected, I get no hints from harper on literate haskell (.lhs) files. Would you accept a PR that adds literate haskell support? I saw it was removed in [grantlemons@c100c6d](https://github.com/grantlemons/harper/commit/c100c6d9955c108526281069340a9579751481a8)
> 
> EDIT: Orthogonal idea: if harper supported analyzing specific ranges of text (does it?), I could configure my neovim lsp client to send only ranges with the `@spell` capture group to harper. Then harper wouldn't have to do any parsing at all, just grammar checking. And it would "support" all languages that are sufficiently supported by my client.

It shouldn't be too difficult to support literate haskell, I just took it out because I wasn't sure how our comment parsing would mesh with how `tree-sitter-haskell` views literate haskell (since it's effectively swapped). I didn't actually test though, so there might be an easy fix already.

As for the `@spell` capture group, that's a bit of a bandaid, and I don't think we really want to add any specific support for things like that, but it sounds feasable just on your end. Best of luck!

**Comment by grantlemons** [comment]
Turns out this is harder than expected to implement, since there is no apparent tree sitter support for literate haskell (correct me if this is not the case).

We are still able to create a custom masker, which I have started on, but it will require some thought since our CommentParser wasn't designed for literate programming.

Thinking about it now, it would probably be smarter to include it as a parser in `harper_core/src/parsers/` instead of `harper_comments/src/comment_parser.rs`.

**Comment by andremarianiello** [comment]
I have been using a custom literate Haskell tree sitter parser. I haven't made it public but if you want to try it out I'm happy to release it. It's super simple because it only separates comments from code, relying on sub parser injections for Haskell and markdown parsing. I don't know if Harper is set up to work with injected parsers, but that probably isn't required.

**Comment by grantlemons** [comment]
Thanks for the offer, but we have a pretty well flushed-out masking system we can use for this, so implementing it wasn't too hard.

**Comment by andremarianiello** [comment]
Cool! Is there a way to configure harper-ls to analyzing everything so that I can do my own masking of the returned diagnostics on the client side? I am just curious if that is an option in case harper has trouble with multiple levels of code+comment nesting.

**Comment by grantlemons** [comment]
I think your closest bet would be to have harper just consider the whole document to be markdown, which would be accomplished by your lsp client sending a language_id of `markdown`. Not sure if that's possible, but it probably is.

**Comment by andremarianiello** [comment]
Yup, I ended up doing this. Change the language id to `markdown` and then filtering diagnostics on the client side based on the presence of the `@spell` capture group. Works great!

**Comment by elijah-potter** [comment]
I'm glad it worked out! What an interesting solution indeed.
